src/v/cloud_storage/manifest.cc

```diff
@@ -0,0 +1,442 @@
+// and manifests will have the same prefix we will be able to do around
+// 3000-5000 req/sec. AWS doc mentions that having only two prefix
+// characters should be enough for most workloads
+// (https://aws.amazon.com/blogs/aws/amazon-s3-performance-tips-tricks-seattle-hiring-event/)
+// We're using eight because it's free and because AWS S3 is not the only
+// backend and other S3 API implementations might benefit from that.
+
+static remote_manifest_path generate_partition_manifest_path(
+  const model::ntp& ntp, model::revision_id rev) {
+    // NOTE: the idea here is to split all possible hash values into
+    // 16 bins. Every bin should have lowest 28-bits set to 0.
+    // As result, for segment names all prefixes are possible, but
+    // for manifests, only 0x00000000, 0x10000000, ... 0xf0000000
+    // are used. This will allow us to quickly find all manifests
+    // that S3 bucket contains.
+    constexpr uint32_t bitmask = 0xF0000000;
+    auto path = ssx::sformat("{}_{}", ntp.path(), rev());
+    uint32_t hash = bitmask & xxhash_32(path.data(), path.size());
+    return remote_manifest_path(
+      fmt::format("{:08x}/meta/{}_{}/manifest.json", hash, ntp.path(), rev()));
+}
+
+remote_manifest_path manifest::get_manifest_path() const {
+    return generate_partition_manifest_path(_ntp, _rev);
+}
+
+remote_segment_path
+manifest::get_remote_segment_path(const segment_name& name) const {
+    auto path = ssx::sformat("{}_{}/{}", _ntp.path(), _rev(), name());
+    uint32_t hash = xxhash_32(path.data(), path.size());
+    return remote_segment_path(fmt::format("{:08x}/{}", hash, path));
+}
+
+const model::ntp& manifest::get_ntp() const { return _ntp; }
+
+const model::offset manifest::get_last_offset() const { return _last_offset; }
+
+model::revision_id manifest::get_revision_id() const { return _rev; }
+
+manifest::const_iterator manifest::begin() const { return _segments.begin(); }
+
+manifest::const_iterator manifest::end() const { return _segments.end(); }
+
+size_t manifest::size() const { return _segments.size(); }
+
+bool manifest::contains(const segment_name& obj) const {
+    return _segments.count(obj) != 0;
+}
+
+bool manifest::add(const segment_name& key, const segment_meta& meta) {
+    auto [it, ok] = _segments.insert(std::make_pair(key, meta));
+    _last_offset = std::max(meta.committed_offset, _last_offset);
+    return ok;
+}
+
+const manifest::segment_meta* manifest::get(const segment_name& key) const {
+    auto it = _segments.find(key);
+    if (it == _segments.end()) {
+        return nullptr;
+    }
+    return &it->second;
+}
+
+std::insert_iterator<manifest::segment_map> manifest::get_insert_iterator() {
+    return std::inserter(_segments, _segments.begin());
+}
+
+manifest manifest::difference(const manifest& remote_set) const {
+    if (_ntp != remote_set._ntp && _rev != remote_set._rev) {
+        throw std::logic_error(fmt_with_ctx(
```

Comment:
beautiful.

Reply:
nit: do we want to have our own exceptions in the cloud_storage namespace. 

also, does it make sense for `redpanda` to continue? if it cannot reconcile, then what fixes this problem? a retry? 

hard to tell if this should be a fatal error instead.